plugin/core/promise.py

```diff
@@ -1,14 +1,24 @@
-from .typing import Any, Callable, List, Tuple, Optional
+from .typing import Callable, Generic, List, Optional, Protocol, Tuple, TypeVar, Union
 import functools
 import sublime
 import threading
 
-ResolveFunc = Callable[..., None]  # Optional argument not supported in Callable so using "..."
-FullfillFunc = Callable[[ResolveFunc], None]
-ThenFunc = Callable[[Any], Any]
+T = TypeVar('T')
+T_contra = TypeVar('T_contra', contravariant=True)
```

Comment:
This one is crazy and I don't fully get it but mypy complains that the type used in `ResolveFunc` is expected to be contravariant. I'm not sure why. Googling reveals that it might be a bug in mypy but I haven't found exactly this case in people's reports.

Reply:
Functions are always contravariant in their input arguments from what I understand.

Reply:
But then why something like `Callable[[T], None` doesn't complain about requiring contravariant type?